clientconn.go

```diff
@@ -854,176 +850,258 @@ func (ac *addrConn) errorf(format string, a ...interface{}) {
 	}
 }
 
-// resetTransport recreates a transport to the address for ac.  The old
-// transport will close itself on error or when the clientconn is closed.
-// The created transport must receive initial settings frame from the server.
-// In case that doesn't happen, transportMonitor will kill the newly created
-// transport after connectDeadline has expired.
-// In case there was an error on the transport before the settings frame was
-// received, resetTransport resumes connecting to backends after the one that
-// was previously connected to. In case end of the list is reached, resetTransport
-// backs off until the original deadline.
-// If the DialOption WithWaitForHandshake was set, resetTrasport returns
-// successfully only after server settings are received.
+// resetTransport makes sure that a healthy ac.transport exists.
 //
-// TODO(bar) make sure all state transitions are valid.
-func (ac *addrConn) resetTransport() error {
-	ac.mu.Lock()
-	if ac.state == connectivity.Shutdown {
-		ac.mu.Unlock()
-		return errConnClosing
-	}
-	if ac.ready != nil {
-		close(ac.ready)
-		ac.ready = nil
-	}
-	ac.transport = nil
-	ridx := ac.reconnectIdx
-	ac.mu.Unlock()
-	ac.cc.mu.RLock()
-	ac.dopts.copts.KeepaliveParams = ac.cc.mkp
-	ac.cc.mu.RUnlock()
-	var backoffDeadline, connectDeadline time.Time
-	for connectRetryNum := 0; ; connectRetryNum++ {
+// The transport will close when it encounters an error, or on GOAWAY,
+// or on deadline waiting for handshake, or when the clientconn is
+// closed. Each iteration creating a new transport will try a different
+// address that the resolver resolved to, until it has tried all
+// addresses. Once it has tried all addresses, it will re-resolve to get
+// a new address list. Once a successful handshake has been
+// received, the list is re-resolved and the next reset attempt will
+// try from the beginning.
+//
+// This method has backoff built in. The backoff amount starts at 0 and
+// increases each time resolution occurs (addresses are exhausted). The
+// backoff amount is reset to 0 each time a handshake is received.
+//
+// If the DialOption WithWaitForHandshake was set, resetTrasport returns
+// successfully only after handshake is received.
+func (ac *addrConn) resetTransport(resolveNow bool) {
+	for {
+		// We want to resolve immediately if this is the first in a line of resets, but not on subsequent
+		// resets unless we:
+		//   - Run out of addresses
+		//   - Successfully receive handshake
+		if resolveNow {
+			ac.mu.Lock()
+			ac.cc.resolveNow(resolver.ResolveNowOption{})
+			ac.mu.Unlock()
+		}
+
+		err := ac.nextAddr()
```

Comment:
So, this is only used in the case where `ac.dopts.waitForHandshake` is set. We basically pass a signal in the form of this specific err up to resetTransport that says, "hey, waitForHandshake failed, go ahead and hard-bail out of creating this transport instead of doing another iteration".

Reply:
I'm not sure I'm following.  If the handshake fails, it is not an "error" per our defined guidelines of an error (https://github.com/grpc/grpc-go/blob/master/Documentation/log_levels.md).  Also, please keep `err` within the `if` statement's scope.

Reply:
Done